### ISSUE TITLE
Cachet Integration

### ISSUE THREAD

**Issue by f0o** [QuxLabs/librenms#9]
Most importantly are Incidents on Alerting but if we could work out metric sharing that'd be great as well.

**Comment by MLWALK3R** [comment]
![Image of Yaktocat](http://i.ytimg.com/vi/Zdf7Afgfq8Q/maxresdefault.jpg)

**Comment by f0o** [comment]
Will publish our working draft tomorrow 9:00 UTC.

**Comment by f0o** [comment]
Here's the current WIP:
`includes/alerts/transport.cachet.php`
```php
/* Copyright (C) 2016 Daniel Preussker <preussker@quxlabs.com>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. */

/**
 * Cachet API Transport
 * @author Daniel Preussker <preussker@quxlabs.com>
 * @copyright 2016 QuxLabs UG, LibreNMS
 * @license GPL
 * @package LibreNMS
 * @subpackage Alerts
 */

$url = $opts['url'];
$protocol = array(
    'visibile' => 1,
    'notify'   => false,
);
$protocol['name']             = 'Incident #'.($obj['id'] ? $obj['id'] : $obj['uid']);
$protocol['message']          = 'Automatic incident regarding '.$obj['name'].".\r\n\r\n";
$protocol['status']           = $opts['incident_status'][$obj['severity']];
$protocol['component_id']     = $opts['component_id'][$obj['hostname']];
$protocol['component_status'] = $opts['component_status'][$obj['severity']];

if( $obj['state'] == 0 ) {
    $protocol['status']           = 4;
    $protocol['component_status'] = 1;
    $protocol['message']         .= 'State changed: Resolved'."\r\n\r\n";
}
elseif( $obj['state'] == 2 ) {
    $protocol['status']           = 3;
    $protocol['component_status'] = 2;
    $protocol['message']         .= 'State changed: Acknowledged'."\r\n\r\n";
}
elseif( $obj['state'] == 3 || $obj['state'] == 4 ) {
   echo 'Skipping worsen/betters'.PHP_EOL;
   return true;
}

if( !isset($protocol['component_id']) ) {
    echo 'No component associated, Skipping.'.PHP_EOL;
    return true;
}
elseif( !isset($protocol['component_status']) || !isset($protocol['status']) ) {
    echo 'No status associated, Skipping.'.PHP_EOL;
    return true;
}
elseif( !empty($opts['relevant_alerts']) && !in_array($obj['name'],$opts['relevant_alerts']) ) {
    echo 'Alert irrelevant, Skipping.'.PHP_EOL;
    return true;
}

$curl = curl_init();
curl_setopt($curl, CURLOPT_URL, $url );
curl_setopt($curl, CURLOPT_RETURNTRANSFER, 1);
curl_setopt($curl, CURLOPT_HTTPHEADER, array('Content-type: application/json','X-Cachet-Token: '.$opts['token']));
curl_setopt($curl, CURLOPT_POSTFIELDS, json_encode($protocol));
$ret = curl_exec($curl);
$code = curl_getinfo($curl, CURLINFO_HTTP_CODE);
if( $code != 200 ) {
    var_dump("Cachet returned Error, retry later",$ret); //FIXME: propper debuging
    return false;
}
return true;
```

Config:
```php
$config['alert']['transports']['cachet'] = array(
    'url'              => 'https://status.example.com/api/v1/incidents',
    'token'            => 'FancyTokenYouGetFromCachet',
    'incident_status'  => array(
        'critical' => 1, //IDs of severity per incident from cachet docs
        'warning'  => 3,
    ),
    'component_id'     => array(
        'device1.fqdn'  => 1, //Device's Hostname to Component-ID mapping, multiple devices per component allowed
        'device2.fqdn'  => 1,
        'device3.fqdn'  => 2,
    ),
    'component_status' => array(
        'critical' => 3, //IDs of severity per component from cachet docs, can vary from the incident's severity
        'warning'  => 2,
    ),
    'relevant_alerts' => array(
        'Devices up/down', //NAME of the rules to care about, should be changed soon I guess
        'Device rebooted',
        'Port status up/down',
        'Port utilisation over threshold',
    )
);
```

**Comment by ToeiRei** [comment]
Some notes for the transport: As soon as it's configured in config.php, it's **already enabled**. Not nice, but working.

component_id config part is about mapping the hostnames LibreNMS knows to the components cachet got. So you need to look them up in the component screen of cachet (click edit and read the URL)

an example would be `'mail.example.com' => 2 //mail system ID in cachet`

**Comment by DerDanilo** [comment]
Why is this not yet pushed into master? This is awesome!
Would also be nice to show some nice metric graphs with cachet based on ping response via Librenms stats.

**Comment by f0o** [comment]
Hi,

The biggest critic and arguably problem is the mapping between device/s and components.

It's extremely ugly. But we have this transport deployed in some production environments. We dont use cachet for graphs but just include librenms' rrd graphs instead.

- f0o

**Comment by ToeiRei** [comment]
What about doing our own status page?

**Comment by DerDanilo** [comment]
_"The biggest critic and arguably problem is the mapping between device/s and components."_
Can't we get LibreNMS services and map those to the components rather than devices itself?
Devices itself should mostly control a group of components, when speaking of several services running on one device. This makes more sense to me.

How do you put them into the status page? Via the Header html option?
I would place them this way as well, but how to do it properly?

**Comment by f0o** [comment]
Actually that's not very far fetched. There is a very simplistic status page already in upstream.

I guess we can easily take that and enhance it.

**Comment by DerDanilo** [comment]
That would be truely awesome! :)

**Comment by f0o** [comment]
@DerDanilo yeah we added the SVGs in the header part in cachet. It worked good enough but ofc isn't as fancy as the cachet graphs.

**Comment by DerDanilo** [comment]
For now I am happy if the device state is mapped properly to the page so my customers know what's going on.

**Comment by ToeiRei** [comment]
Just thinking of a typical example: User needs to know if "Mail" works - which is a bunch of services starting from the spamfilter,. smtp and maybe imap daemon. Defining those groups and getting the status of every device would do the trick.

If one component has problems, we can show a red signal.

**Comment by f0o** [comment]
Please bare in mind that LibreNMS is still very device centric.
It's not well suited for application level monitoring and alerting.

The alert states are bound to the device with no notion of components within that device. So whether your mta died or your imap died will both lead in showing the device as alerted and thus flip the same cachet component because it cant differ between mta/imap.

**Comment by f0o** [comment]
But about the reliability, we deployed it to a mid-size ISP which uses it for ~200 switches and routers and haven't heard any major complications. Initial fiddling with config of course.

**Comment by DerDanilo** [comment]
I don't want to map all those services. Cachet is for the customer faced "services". So that means for me "Mailserver". An if I have a simple mapping with an LNMS service check that checks if it can reach the IMAP server (whatever it is), that would be enough.

**Comment by f0o** [comment]
That's entirely possible and feasible :)

**Comment by ToeiRei** [comment]
I mean, if the SMTP dies, the mail service is 'broken' from the customers point. Same for the IMAP service or anything in between. (Read: "I do not get my mails")

**Comment by DerDanilo** [comment]
Right, but if we can actually map a service ID in librenms to the comp ID in cachet that would be awesome.
If a device hits, it then can trigger a component group to show a warning for all of them, even before a warning hits for a service.

**Comment by DerDanilo** [comment]
Can one please give an example of a full generated LibreNMS API call to cachet?
I want to test if it is actually working, before I put more time into finding why it does not seem to trigger right now.

**Comment by f0o** [comment]
Firstly you can't map a service ID to an alert because the alerts are device centric.

As for the API call, it sends $protocol as JSON to the v1/incidents endpoint.

You can var_dump it and investigate it without calling the whole curl block to dryrun the transport

**Comment by DerDanilo** [comment]
Is the Cachet transport supposed to appear in LibreNMS GUI?
Right now I can't get it to work and don't know how to "var_dump" the output. I am no programmer, need some help here. :)

**Comment by f0o** [comment]
First things first;

> Is the Cachet transport supposed to appear in LibreNMS GUI?
No, that requires some other code parts. If there's interest in this and a nice approach on how to handle mapping between DeviceIDs and ComponentIDs then we're more than happy to make PR upstream including a GUI Config.

> Right now it only seems to trigger once I "Acknowledge" the error message in LibreNMS, before that it does NOT trigger.
That's odd, can you paste the config part please?

> Does the plugin automatically set the component to "operational" once it is available again?
Yes, when the alert is cleared it will be set to operational again.

**Comment by shibumi** [comment]
Any update on this? Why is this not merged yet? Thats awesome

**Comment by dermotw** [comment]
Just came across this and am intrigued. We have a fairly large network, so I'm not too excited about manually editing the config file to add the device -> component mappings or, worse, letting people who aren't me edit the config. I've thought of a couple of workarounds though:

1. Entering the name of the component that the device's status should affect into its notes and then using $obj['notes'] as the basis for the component mapping. This would allow us to use the LibreNMS GUI to assign devices to components. As a bonus, with some extra work in transport.cachet.php, we could assign devices to multiple components. Useful in our case since a given network device could be involved in providing/carrying two or more services.

2. I'm not sure how this would work but an alternative would be to map groups to component IDs in config.php and then somehow work out what group(s) a given device belongs to in transport.cachet.php.

I'm going to think about 2 and might do some work on 1 since it seems straightforward enough.

**Comment by DerDanilo** [comment]
Can someone write a script that just changes the state of components without posting incidents?
